N°6901 - Enable tracking of iTop active sessions (monitoring)

odain-cbd opened the pull request:
### Goal

Current PR aims at being able at any time to know the amount of user currently active on iTop.
Design

For that purpose a dedicated iTop session handler extends default one. it keeps default implementation by calling each parent method first.
and then a specific iTop implementation is triggered to have local session files in data/sessions folder.

**session file content**
each session file is either empty when iTop session is made without any user authentified (status.php for example).
otherwise it contains below data:

-  login_mode: authentication mode (among UI classical 'form', 'token', 'ldap', sso called 'hybrid-auth', .....)
- user_id: logged in user ID
- creation_time: timestamp of the PHP session creation (used to compute session elapsed active time in seconds)
- context: iTop session context tags are aggregated with ':' separators (empty when simple login or pick among ContextTag values like REST/Synchro/....)

Examples:

- user UI access:

`{"login_mode":"form","user_id":"1","creation_time":1699371169,"context":""}`

- Rest API call (via token)

`{"login_mode":"token","user_id":"4","creation_time":1699373876,"context":"REST\/JSON"}`

- Synchro API call (via token as well)
`{"login_mode":"token","user_id":"6","creation_time":1699373876,"context":"Synchro"}`

**Cleanup**
Garbage collecting of all local session files (in data/sessions folder) is made via a dedicated Background task.

odain-cbd committed:
N°6901 - Add itop session monitoring metric

odain-cbd committed:
N°6901 - create itop session folder when missing

odain-cbd committed:
N°6901 - GC fix + add session_mode in session files

odain-cbd committed:
fix ci

odain-cbd committed:
fix missing classes in get_declared_classes result

odain-cbd committed:
N°6901-fix namespace + add context in itop local session

odain-cbd committed:
N°6901 - fix GarbageCollector use in cron + extract session file content method

odain-cbd committed:
N°6901- cleanup unahtenticated sessions immediately

odain-cbd committed:
6901-add phpdoc + cleanup unahtentified sessions after 1 minute

odain-cbd committed:
N°6901-fix context

odain-cbd committed:
N°6901 - fix broken cron

odain-cbd committed:
N°6901 - fix session cleanup background process

odain-cbd committed:
N°6901 - include session creation_time in session files to expose session elapsed time + fix session_regenerate + tests

Molkobain commented on sources/Application/Helper/iTopSessionHandler.php:
Move this to `sources/SessionTracker/SessionHandler.php`

Molkobain commented on core/cmdbsource.class.inc.php:
```diff
@@ -518,7 +518,12 @@ public static function Quote($value, $bAlways = false, $cQuoteStyle = "'")
 		{
 			/** @noinspection NullPointerExceptionInspection this shouldn't be called with un-init DB */
-			$value = $cQuoteStyle.DbConnectionWrapper::GetDbConnection()->real_escape_string($value).$cQuoteStyle;
+			$mysqli = DbConnectionWrapper::GetDbConnection();
```
As this change is not really necessary for the PR, but fixed a test case, we'll see if we keep it during PR review.

odain-cbd replied:
just reverted

Molkobain committed:
Remove require as background task is automatically detected

Molkobain committed:
Refactor to match code conventions

Molkobain committed:
Limit indentation hell

Molkobain committed:
Use utils::GetDataPath() instead of hard-coded path

Molkobain commented on sources/Application/Helper/iTopSessionGarbageCollector.php:
Move this to `sources/SessionTracker/SessionTrackerGC.php`

- No need to prefix with iTop
- Other GC task are suffixed with "GC" instead of "GarbageCollector"

Molkobain commented on lib/composer/autoload_classmap.php:
Mind to redump autoloaders after changes below.

Molkobain commented on sources/Application/Helper/iTopSessionHandler.php:
```diff
@@ -0,0 +1,181 @@
+ * @package Combodo\iTop\Application\Helper
+ * @since 3.1.1 3.2.0
+ */
+class iTopSessionHandler extends \SessionHandler
```
Put overrided methods first, then custom methods.

Molkobain commented on sources/Application/Helper/iTopSessionHandler.php:
```diff
@@ -0,0 +1,181 @@
+					'login_mode' => Session::Get('login_mode'),
+					'user_id' => $sUserId,
+					'creation_time' => $iCreationTime,
+					'context' => implode(":", ContextTag::GetStack())
```
Don't use ":" as it is already used in some context tags such as cron tasks, portal instances, synchro DS, ...
Discussed this morning, using "|" seems safe.

Molkobain commented on sources/Application/Helper/iTopSessionHandler.php:
```diff
@@ -0,0 +1,181 @@
+			@mkdir(utils::GetDataPath() . "sessions");
+		}
+
+		return glob(utils::GetDataPath() . "sessions/session_**");
```
```suggestion
		return glob(utils::GetDataPath() . "sessions/session_*");
```

Molkobain commented on webservices/cron.php:
```diff
@@ -250,7 +250,7 @@ function CronExec($oP, $bVerbose, $bDebug=false)
 					$sLastRunDate = $oTask->Get('latest_run_date');
 					$sNextRunDate = $oTask->Get('next_run_date');
-					$oP->p(sprintf('| %1$-25.25s | %2$-7s | %3$-19s | %4$-19s |', $sTaskName, $sStatus, $sLastRunDate, $sNextRunDate));
+					$oP->p(sprintf('| %1$-50.50s | %2$-7s | %3$-19s | %4$-19s |', $sTaskName, $sStatus, $sLastRunDate, $sNextRunDate));
```
```suggestion
					$oP->p(sprintf('| %1$-25.25s | %2$-7s | %3$-19s | %4$-19s |', $sTaskName, $sStatus, $sLastRunDate, $sNextRunDate));
```

My bad

Molkobain reviewed (changes requested):
Discussed this morning with Olivier and Eric:
- I commited some changes
- Olivier will do the others

Molkobain committed:
Update webservices/cron.php

Molkobain commented on sources/Application/Helper/iTopSessionHandler.php:
```diff
@@ -0,0 +1,181 @@
+		$bRes = parent::open($save_path, $session_name);
+
+		$session_id = session_id();
+		IssueLog::Debug("open($session_id)");
```
All logs need to be more verbose: At least tell that you are doing something about the sessions. Better, add a channel and context data.

Molkobain commented:
_Looks like my notes for this morning review got sent with the ones I just added now. You might have received them by mail, I deleted the old ones, just mind the ones still on the PR_

odain-cbd committed:
N°6901 - PR review feedbacks with G. and E.

odain-cbd commented:
> Discussed this morning with Olivier and Eric:
> 
>     * I commited some changes
> 
>     * Olivier will do the others

just made all corrections based on your feedbacks. thanks

Molkobain commented on core/log.class.inc.php:
```diff
@@ -576,6 +576,7 @@ class LogChannels
 	public const DATATABLE = 'Datatable';
 
 	public const DEADLOCK = 'DeadLock';
+	public const SESSIONTRACKER = 'SessionTracker';
```
```suggestion
	/**
	 * @var string Everything related to PHP sessions tracking
	 * @since 3.1.1 3.2.0 N°6901
	 */
	public const SESSIONTRACKER = 'SessionTracker';
```

Molkobain commented on sources/SessionTracker/SessionHandler.php:
```diff
@@ -0,0 +1,193 @@
+use utils;
+
+/**
+ * Class iTopSessionGarbageCollector
```
```suggestion
 * Class SessionHandler
```

Molkobain commented on sources/SessionTracker/SessionHandler.php:
```diff
@@ -0,0 +1,193 @@
+ * Class iTopSessionGarbageCollector
+ *
+ * @author Olivier Dain <olivier.dain@combodo.com>
+ * @package Combodo\iTop\Application\Helper
```
```suggestion
 * @package Combodo\iTop\SessionTracker
```

Molkobain commented on sources/SessionTracker/SessionGC.php:
```diff
@@ -0,0 +1,32 @@
+ * Class SessionGC
+ *
+ * @author Olivier Dain <olivier.dain@combodo.com>
+ * @package Combodo\iTop\Application\Helper
```
```suggestion
 * @package Combodo\iTop\SessionTracker
```

Molkobain commented on sources/SessionTracker/SessionHandler.php:
```diff
@@ -0,0 +1,193 @@
+	{
+		$bRes = parent::write($session_id, $data);
+
+		IssueLog::Debug("gc", \LogChannels::SESSIONTRACKER, [
```
```suggestion
		IssueLog::Debug("write", \LogChannels::SESSIONTRACKER, [
```

Molkobain committed:
Apply suggestions from code review

Molkobain commented:
Great thanks! I'll set up a meeting within engineering for the follow-up.

Molkobain commented on sources/SessionTracker/SessionHandler.php:
```diff
@@ -0,0 +1,193 @@
+{
+	public function destroy($session_id)
+	{
+		IssueLog::Debug("destroy", \LogChannels::SESSIONTRACKER, [
```
```suggestion
		IssueLog::Debug("Destroy PHP session", \LogChannels::SESSIONTRACKER, [
```

Molkobain commented on sources/SessionTracker/SessionHandler.php:
```diff
@@ -0,0 +1,193 @@
+
+	public function gc($max_lifetime)
+	{
+		IssueLog::Debug("gc", \LogChannels::SESSIONTRACKER, [
```
```suggestion
		IssueLog::Debug("Run PHP sessions garbage collector", \LogChannels::SESSIONTRACKER, [
```

Molkobain commented on sources/SessionTracker/SessionHandler.php:
```diff
@@ -0,0 +1,193 @@
+		$bRes = parent::open($save_path, $session_name);
+
+		$session_id = session_id();
+		IssueLog::Debug("open", \LogChannels::SESSIONTRACKER, [
```
```suggestion
		IssueLog::Debug("Open PHP session", \LogChannels::SESSIONTRACKER, [
```

Molkobain commented on sources/SessionTracker/SessionHandler.php:
```diff
@@ -0,0 +1,193 @@
+	{
+		$bRes = parent::write($session_id, $data);
+
+		IssueLog::Debug("write", \LogChannels::SESSIONTRACKER, [
```
```suggestion
		IssueLog::Debug("Write PHP session", \LogChannels::SESSIONTRACKER, [
```

Molkobain commented on tests/php-unit-tests/unitary-tests/sources/SessionTracker/SessionHandlerTest.php:
```diff
@@ -0,0 +1,299 @@
+	}
+
+	private function CreateUserAndLogIn() : ? string {
+		$_SESSION = [];
```
This might not be necessary anymore. We should discuss with @rquetiez

Molkobain reviewed (commented):
Specifying the log channel is great step forward as it enables us to set the log level for these specific logs.

That being said, I think it would ease reading logs (from all channels) for Support / R&D teams if the messages where a little bit more explicit and say that they are about PHP session.

What do you think @odain-cbd @eespie ?

Molkobain commented on sources/SessionTracker/SessionHandler.php:
```diff
@@ -0,0 +1,193 @@
+					'login_mode' => Session::Get('login_mode'),
+					'user_id' => $sUserId,
+					'creation_time' => $iCreationTime,
+					'context' => implode("|", ContextTag::GetStack())
```
```suggestion
					'contexts' => implode("|", ContextTag::GetStack())
```

Making this plural as it contains all context tags currently set.
@odain-cbd you might need to adjust your monitoring module.

Molkobain committed:
Update sources/SessionTracker/SessionHandler.php

Molkobain commented on sources/SessionTracker/SessionHandler.php:
```diff
@@ -0,0 +1,193 @@
+					'login_mode' => Session::Get('login_mode'),
+					'user_id' => $sUserId,
+					'creation_time' => $iCreationTime,
+					'contexts' => implode("|", ContextTag::GetStack())
```
```suggestion
					'context' => implode("|", ContextTag::GetStack())
```

Previous change was actually stupid as we already use singular everywhere else.

Molkobain committed:
Revert previous commit

odain-cbd committed:
Update sources/SessionTracker/SessionHandler.php

Co-authored-by: Molkobain <lajarige.guillaume@free.fr>

odain-cbd committed:
Update sources/SessionTracker/SessionHandler.php

Co-authored-by: Molkobain <lajarige.guillaume@free.fr>

odain-cbd committed:
Update sources/SessionTracker/SessionHandler.php

Co-authored-by: Molkobain <lajarige.guillaume@free.fr>

odain-cbd committed:
Update sources/SessionTracker/SessionHandler.php

Co-authored-by: Molkobain <lajarige.guillaume@free.fr>

odain-cbd committed:
specify proper SessionHandler signatures to open/gc/write/destroy

odain-cbd committed:
fix ci detectStaticPollutionTest.testDetectPolluters

odain-cbd committed:
fix ci detectStaticPollutionTest.testDetectPolluters

Molkobain commented:
~@odain-cbd they is an issue with the PR, there now are 29 commits and some changed files that should not be there.
Let's check together when you have a moment.~

odain-cbd committed:
fix ContextTag by replacing any AddContext call

odain-cbd committed:
PHP doc

odain-cbd committed:
fix ci by adding ReturnTypeWillChange on inherited APIs from SessionHandler

Molkobain commented:
Discussed IRL with Alexandre, Olivier and Romain:
- Planned for iTop 3.1.1 / 3.2.0
- Add random session GC trigger on session opening to ensure files are cleared even without a CRON running. See what is done in `\privUITransactionFile::CleanupOldTransactions()`
- Session handler override will be conditioned to a config. param. disabled by default.
```php
		'session.enable_custom_session_handler' => [
			'type' => 'bool',
			'description' => 'Enable custom handler to have PHP sessions logs (opened, destroyed, ...) and generate additional session files for monitoring purposes',
			'default' => false,
			'value' => false,
			'source_of_value' => '',
			'show_in_conf_sample' => false,
		],
```

odain-cbd committed:
fix ci SessionHandler API notices

odain-cbd committed:
make session tracking configurable and disabled by default + run it live to have minimal cleanup even without cron installed

odain-cbd committed:
fix notice

rquetiez commented on core/config.class.inc.php:
```diff
@@ -1193,6 +1193,30 @@ class Config
 			'source_of_value' => '',
 			'show_in_conf_sample' => false,
 		],
+		'sessions_tracking_enabled' => [
```
It is more common to use a dot between a feature name and the various settings.
=> 'sessions_tracking.enabled'

odain-cbd replied:
yes it could be nicer. but this is not compliant with other existing variables like _transactions_gc_threshold_...

i could also have added an array instead (my first choice before seeing transaction* vars):
```
'sessions_tracking' =>
[ 'enabled' => true, 'gc_collector' => 'xx', 'gc_timelimit' => 'YY']
```

rquetiez commented on core/config.class.inc.php:
```diff
@@ -1193,6 +1193,30 @@ class Config
+		],
+		'sessions_tracking_gc_threshold' => [
+			'type' => 'integer',
+			'description'         => 'probability in percent for the garbage collector to be triggered (100 mean always)',
```
What does that mean? What happens when the garbage collector is triggered?

odain-cbd replied:
it is the probability the gc is triggered in fact at each Session.start(). nothing new as I relied on transaction existing cleanup logic

rquetiez commented on sources/SessionTracker/SessionHandler.php:
```diff
@@ -0,0 +1,234 @@
+ * @package Combodo\iTop\SessionTracker
+ * @since 3.1.1 3.2.0 N°6901
+ */
+class SessionHandler extends \SessionHandler
```
This is a bit mysterious. I had to focus on the namespace to figure out what was the purpose of such a class having the same name as its parent class.
I suggest adding a comment like "Tracks active user sessions in files"
Possibly also name the class like "FileTrackedSession"

odain-cbd replied:
good point. still name should end with SessionHandler at least.

defaut PHP SessionHandler already relies on files. this new iTop SessionHandler keeps on calling his parent first. and afterwards creates files apart in itop folders (/data/...) with very few data.

what about FileActiveSessionTrackerSessionHandler? or ActiveSessionTrackerSessionHandler? seems quite long, no?

odain-cbd replied:
add phpdoc only but no renaming

Hipska replied:
What about `TrackingSessionHandler`?

Molkobain replied:
I asked Olivier to named it like that.

For information, it is very common (and a good practice) to have a class with the same name as the one it is extending, the namespace is here to differenciate them.

To ease the understanding, the usual convention is to do something like this (I'm not asking Olivier to do it, just we spread awereness):

```php
namespace A\B;

use SessionHandler as BaseSessionHandler; // This makes an alias for this class to be used in this file

class SessionHandler extends BaseSessionHandler
{
    [...]
}
```

rquetiez commented on tests/php-unit-tests/unitary-tests/sources/SessionTracker/SessionHandlerTest.php:
```diff
@@ -0,0 +1,303 @@
+		$this->assertNotNull($sContent);
+		$aJson = json_decode($sContent, true);
+		$this->assertNotEquals(false, $aJson, $sContent);
+		$this->assertEquals($sUserId, $aJson['user_id'] ?? '', $sContent);
```
In case the test fails, the message will be cryptic => $sContent
I would rather have something like "Session content should have the user_id entry, found $sContent"

rquetiez replied:
Same comment for most of the assertions above and below.

rquetiez commented on tests/php-unit-tests/unitary-tests/sources/SessionTracker/SessionHandlerTest.php:
```diff
@@ -0,0 +1,303 @@
+		$this->assertEquals($sContent, $sNewContent, $sNewContent);
+	}
+
+	public function testTouchSessionFile_UserLoggedInWithImpersonation(){
```
The first part (until UserRights::Impersonate) seems 100% redundant with the preceding test. Either keep only this test (prefered for speed of tests), or remove all the assertions of the first part.

odain-cbd replied:
merged some tests and add comments in case of failures

rquetiez commented on tests/php-unit-tests/unitary-tests/sources/SessionTracker/SessionHandlerTest.php:
```diff
@@ -0,0 +1,303 @@
+		$this->assertEquals($sFirstContent, $sNewContent, $sNewContent);
+	}
+
+	public function testGenerateSessionContentWithSessionContextChange(){
```
The first part is 100% identical to the previous tests.
Please Keep the complete version and rename it to testGenerateSessionContent, for the sake of clarity.

rquetiez commented on tests/php-unit-tests/unitary-tests/sources/SessionTracker/SessionHandlerTest.php:
```diff
@@ -0,0 +1,303 @@
+
+		$aFoundSessionFiles = $oSessionHandler->list_session_files();
+		foreach ($aFoundSessionFiles as $sFile){
+			$this->assertTrue(is_file($sFile));
```
$this->assertTrue(is_file($sFile), 'list_session_files is should return only existing files. Not found: '.$sFile);

rquetiez commented on tests/php-unit-tests/unitary-tests/sources/SessionTracker/SessionHandlerTest.php:
```diff
@@ -0,0 +1,303 @@
+		$oSessionHandler->gc_with_time_limit(-1);
+		$this->assertEquals([], $oSessionHandler->list_session_files());
+
+		for($i=0; $i<=1; $i++) {
```
Please use a constant or variable to help the reader understanding the relation between this loop and the assertion on line 264.
Suggestion :
for ($i = 0 ; $i < FAKE_FILE_COUNT ; $i++)

rquetiez commented on tests/php-unit-tests/unitary-tests/sources/SessionTracker/SessionHandlerTest.php:
```diff
@@ -0,0 +1,303 @@
+		$oSessionHandler->gc_with_time_limit(-1);
+		$this->assertEquals([], $oSessionHandler->list_session_files());
+
+		for($i=0; $i<=1; $i++) {
```
for ($i = 0 ; $i < FAKE_FILE_COUNT ; $i++)

rquetiez commented on tests/php-unit-tests/unitary-tests/sources/SessionTracker/SessionHandlerTest.php:
```diff
@@ -0,0 +1,303 @@
+	/**
+	 * @dataProvider GgcWithTimeLimit_EmptyFile
+	 */
+	public function testGgcWithTimeLimit_EmptyFile($iTimeLimit, $iExpectedProcessed) {
```
What is the difference between an empty file and a file containing the word "fake"?
I wonder if this test could be merged with the previous one.

odain-cbd replied:
one test consists in checking empty files are cleaned up. 

the other is a mix where I test cleanup works on a list of existing non empty files and files that do not existing (fake should be renaming: non-existing-file).

odain-cbd replied:
empty files are for unauthentified active sessions (when we call status.php). in most cases files are not empty and there are authentication data inside (login_mode/context/user_id).

odain-cbd committed:
php doc

odain-cbd committed:
fix MetaModel->GetConfig null

odain-cbd committed:
conf variables renaming + doc

odain-cbd committed:
PR review: enhance test with Romain feedbacks

rquetiez committed:
(could still be simplified even more)
Add @covers annotations
Cleanup all files created during the test
Rephrased almost all assertion messages in order the reflect the API expected behavior or at least the fact that the test relies on some other APIs to work fine.

rquetiez committed:
Latest changes before squashing...

rquetiez reviewed (approved):
Reviewed the tests
